[NTOS:CM] Implement NtNotifyChangeMultipleKeys

xmine64 opened the pull request:
## Purpose

This PR provides a series of commits implementing NtNotifyChangeMultipleKeys

JIRA issue: [CORE-11865](https://jira.reactos.org/browse/CORE-11865)

Related documentations:
https://learn.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntnotifychangemultiplekeys
https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-zwnotifychangekey
https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/using-nt-and-zw-versions-of-the-native-system-services-routines

The implementation is inspired by my analysis of Windows kernel using a disassembler and behavior of the API test on Windows 2003, Windows 7, Windows 10 and Windows 11, but there are things that has been done differently, like using an array for storing Subordinate objects, since it supposed to support more than one object but in Windows only one subordinate object is supported.

## Proposed changes
* ~Size of the output buffer in `ListTests` in KM test is increased, since it wasn't enough to store names of the tests after adding the new test.~ (Fixed in https://github.com/reactos/reactos/commit/21d02b82668477ddafd24f204e3ecb3a4740a7dd)
* Previously `CmpInitializeHive` used to initialize `CMHive->NotifyList` by setting its Flink and Blink to NULL, I updated this to use `InitializeListHead` so I can insert NotifyBlocks to the hive.
* An implementation for `CmpReportNotify` is added, it works by searching CMHive->NotifyList for a NotifyBlock that's attached to the given KCB. I couldn't understand how does this works on Windows but this is a solution I could came by. I tried checking the KeyBody for NotifyBlock but it didn't work.
* An implementation for `CmpFlushNotify`, this releases all the resources used by the NotifyBlock, and system calls it when it wants to free a registry key handle
* Helper functions added to support the functionality: `CmpIsKcbSubKey`, `CmpPostBlockFreeSubordinates`, `CmpNotifyPostBlock`, `CmpInsertNotifyBlock`, `CmpInsertPostBlock`, `CmpInsertSubPostBlock`
* `CmpCloseKeyObject` is updated to call `CmpFlushNotify` instead of asserting when it finds a NotifyBlock
* A new internal structure called `CM_POST_BLOCK` is added

The whole thing works by inserting a NotifyBlock to the KeyBody when users calls `NtNotifyChangeMultipleKeys`. NotifyBlock represents the "notification session" for a key handle. Also another internal structure called `CM_POST_BLOCK` is allocated and attached to the NotifyBlock->PostList which represents a "notification listener". With every change in the registry, our CM functions call `CmpReportNotify`, signals each PostBlock on each NotifyBlock for the updated registry key. The signal can be sent using an "Event" object, a "KAPC", or a "QUEUE_WORK_ITEM".

## TODO

- [x] apitests: user-mode
- - [x] Synchronous
- - [x] Asynchronous
- - [x] Multiple keys and watching subtree
- [x] apitests: kernel-mode
- [x] User-mode implementation
- - [x] Synchronous
- - [x] Event-based Asynchronous
- - [x] APC-based Asynchronous
- - [x] Watch subtree
- - [x] Multiple keys
- [x] Kernel-mode `WORK_QUEUE_ITEM`-based asynchronous
- [x] Sending change notifications
- [x] Return IoStatusBlock for user-mode callers
- [ ] Return name of changed key in Buffer (unimplemented on Windows)

## Testbot runs (Filled in by Devs)

- [ ] KVM x86:
- [ ] KVM x64:

HBelusca commented:
Note that your PR will require writing apitests.

HBelusca commented on ntoskrnl/config/cmnotify.c:
```diff
@@ -14,6 +14,77 @@
 
 /* FUNCTIONS *****************************************************************/
 
+PCM_NOTIFY_BLOCK
+NTAPI
+CmpAllocateNotifyBlock(IN CM_NOTIFY_MODE NotifyMode,
+                       IN PCM_KEY_BODY Key,
+                       IN PCM_NOTIFY_BLOCK MasterBlock OPTIONAL)
```
For new functions, please use SAL2 annotations. Here's an example:
```suggestion
PCM_NOTIFY_BLOCK
NTAPI
CmpAllocateNotifyBlock(
    _In_ CM_NOTIFY_MODE NotifyMode,
    _In_ IN PCM_KEY_BODY Key,
    _In_opt_ PCM_NOTIFY_BLOCK MasterBlock)
```

HBelusca commented on ntoskrnl/include/internal/cm.h:
```diff
@@ -332,6 +340,10 @@ typedef struct _CM_NOTIFY_BLOCK
     ULONG WatchTree:30;
     ULONG NotifyPending:31;
+
+    CM_NOTIFY_MODE NotifyMode;
```
Either the following 3 members aren't actually needed
(they are absent from the official symbols, see e.g.:
https://www.vergiliusproject.com/kernels/x86/server-2003/sp2/_CM_NOTIFY_BLOCK
https://www.vergiliusproject.com/kernels/x64/windows-10/22h2/_CM_NOTIFY_BLOCK
https://www.vergiliusproject.com/kernels/x64/windows-11/24h2/_CM_NOTIFY_BLOCK
),
or this means that the notify_block structure is actually being wrap into yet another internal structure, that contains those extra members.

xmine64 replied:
I think it might be actually wrapped inside another structure. It was referred as "Post Block" instead of "Notify Block".

xmine64 replied:
I just found that "Post Block" is a new internal structure completely different from "Notify Block" and seems like `CM_POST_BLOCK` and `CmpAllocatePostBlock` are not documented anywhere.

xmine64 commented:
> Note that your PR will require writing apitests.

There's a winetest on this and I thought that would be enough, but Ok I'll add a apittest too.

https://github.com/reactos/reactos/blob/2c2cdfddd3c7189cd9639b92f2204427d0aa1f8a/modules/rostests/winetests/ntdll/reg.c#L1893-L1920

xmine64 commented:
I'm restarting from scratch, as I felt I've been doing it in a wrong way.
I'll start with a simplified version of this function with missing features, and I'll add support for more scenarios step-by-step.
This PR is going to be a bit larger than just implementing a single function, so I'm ok with breaking it into smaller PRs if it's needed.

xmine64 commented:
Now I confirmed when NtNotifyChangeMultipleKeys called in synchronous mode, it waits for a change to happen, but I need to implement support for reporting notifications so we can signal it when the key changed.
Previously the apitest I wrote was failing because NtNotifyChangeMultipleKeys returned too soon without actually waiting, now it fails because it never returns.
Also now calling CmpCloseKeyHandle on keys where NtNotifyChangeMultipleKeys called on causes a kernel panic because the routine to free NotifyBlock is not implemented yet.

xmine64 commented:
Now I have a basic working implementation of NtNotifyChangeKey for synchronous and asynchronous calls from user-mode:
![image](https://github.com/user-attachments/assets/7f6804ac-1e9f-4386-9f77-fbd5b3e21b94)

For remaining works:
* I need to add API test for APC mode calls from user-mode, but I don't know anything about APC calls on Win32.
* Also, the current implementation ignores WatchTree (reporting notifications for subkeys) and SlaveObjects (A secondary registry key to watch, alongside the master key). I didn't write API tests for these scenarios too, but I'll add soon.

* For kernel-mode asynchronous calls:
Based on documentations provided by Microsoft [here](https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/using-nt-and-zw-versions-of-the-native-system-services-routines):
> For a call from kernel mode, ApcRoutine points to a [WORK_QUEUE_ITEM](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/ns-wdm-_work_queue_item) structure, and ApcContext specifies the type of work queue item that is described by the WORK_QUEUE_ITEM structure.

I need to use WorkQueueItem for asynchronous kernel-mode calls. this would also make our btrfs driver work better as it depends on ZwNotifyChangeMultipleKeys.
I'll add API tests for kernel mode calls too.

xmine64 commented:
Update:
Now I have API tests and KM tests covering every feature of NtNotifyChangeMultipleKeys, and this is their output on Windows Server 2003:

![image](https://github.com/user-attachments/assets/3f179e66-ae78-4939-a925-fcccaa59763b)

And this is ReactOS's status:

![image](https://github.com/user-attachments/assets/cabd3ee1-9009-43ce-b92f-f4c3c7823df7)

xmine64 commented:
Thanks to all the community members who helped me along the path, now all API tests and KM tests pass and the code is ready to be merged. WINE test doesn't pass but there should be something wrong with the test itself.
This is my first time writing something to be ran in the Windows NT kernel, so review carefully as I'm not confident if I handled everything correctly.

@GeoB99 @HBelusca I'll be waiting for your feedback

GeoB99 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+                goto Failure;
+
+            /* Check if this is a duplicate object */
+            if (LocalSubObjectsKeyBody[i]->KeyControlBlock == KeyObject->KeyControlBlock)
```
I suppose you are checking here if the KCB of the sub key matches with the primary key object to avoid locking it again, in order to prevent deadlocks? Why not just simply check if the sub KCB is different from the main one, and lock the KCB for exclusive use right away?

xmine64 replied:
I guess it would cause a deadlock if a subordinate is pointing the same key as master, but well I added this check because it assumes each provided key is a different key. You're supposed to provide different keys when you want to watch multiple keys. Come to think about this, Windows locks the Hive in this function and also it prevents you from providing a subordinate in the same Hive as the master key, and this should be the reason.

GeoB99 replied:
Please provide a testcase in your KMTEST that indeed Windows throws `STATUS_INVALID_PARAMETER` if the caller submits the same subordinate key as that of the master one.

>I guess it would cause a deadlock

Of course it will cause a deadlock, as KCBs are protected by using a push lock and if those aren't used judiciously then you're gonna freeze the whole system as both threads are waiting on each other.

xmine64 replied:
> Please provide a testcase in your KMTEST that indeed Windows throws STATUS_INVALID_PARAMETER if the caller submits the same subordinate key as that of the master one, if you haven't done it so.

I didn't, I'll do OK. Also Windows returns error even if the subordinate key is in the same Key Hive as the master, so it doesn't have to be the same key. The error code is different on Windows 2000 and Windows 11, where Windows 2000 returns `STATUS_INVALID_PARAMETER` but Windows 11 returns `STATUS_INVALID_OBJECT_NAME`.

xmine64 replied:
The test is provided now but in NTDLL_APITEST instead of KMTEST. I'll mark this as resolved if there's no other feedback.

GeoB99 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+        return STATUS_INVALID_PARAMETER;
+
+    /* Windows doesn't support more than one subordinate */
+    if (Count != 0 && Count != 1)
```
`Count` is a ULONG so if the implementation only supports one subordinate you could just do `Count > 1` if it's above the supported subordinate.

GeoB99 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+        MasterKeyHandle, KEY_NOTIFY, CmpKeyObjectType, PreviousMode, (PVOID *)&KeyObject, NULL);
+
+    if (!NT_SUCCESS(Status))
+        return Status;
```
It would be better for failure paths to print out the failure code with DPRINT1 indicating why the operation failed, like we normally do across all of kernel code.

GeoB99 replied:
`DPRINT1`

xmine64 replied:
Ok then I'll change them to `DPRINT1` soon.

GeoB99 commented on ntoskrnl/config/cmnotify.c:
```diff
@@ -12,25 +12,481 @@
+ */
+BOOLEAN
+NTAPI
+CmpIsKcbSubKey(_In_ PCM_KEY_CONTROL_BLOCK ParentKey, _In_ PCM_KEY_CONTROL_BLOCK SubKey)
```
The parameters must be declared in cascade way.

GeoB99 commented on ntoskrnl/config/cmnotify.c:
```diff
@@ -12,25 +12,481 @@
+    PCM_NOTIFY_BLOCK NotifyBlock;
+    PCMHIVE Hive;
+
+    /* Allocate memory */
+    NotifyBlock = ExAllocatePoolWithTag(NonPagedPool, sizeof(CM_NOTIFY_BLOCK), TAG_CM);
+    if (!NotifyBlock)
+    {
+        *Result = NULL;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    RtlZeroMemory(NotifyBlock, sizeof(CM_NOTIFY_BLOCK));
```
Use `ExAllocatePoolZero` for this and so you've saved yourself from having `RtlZeroMemory` in the code.

GeoB99 reviewed (dismissed):
There's a bunch mix of code change and unnecessary formatting style change in between of which I wonder why you're even touching other unrelated parts. It makes the PR quite harder to review. I appreciate your work but in no way I would accept such a PR like this in this state.

I have posted some technical review questions, I haven't reviewed your tests yet. I will get to those when I can.

xmine64 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -17,24 +17,19 @@ BOOLEAN CmBootAcceptFirstTime = TRUE;
-ReleaseCapturedObjectAttributes(
-    _In_ POBJECT_ATTRIBUTES CapturedObjectAttributes,
-    _In_ KPROCESSOR_MODE AccessMode)
+ReleaseCapturedObjectAttributes(_In_ POBJECT_ATTRIBUTES CapturedObjectAttributes, _In_ KPROCESSOR_MODE AccessMode)
```
I thought I only selected my changed codes when I ran clang-format,I'll revert it asap.

xmine64 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
-                           OUT PVOID Buffer,
-                           IN ULONG Length,
-                           IN BOOLEAN Asynchronous)
+NtNotifyChangeMultipleKeys(
```
I wanted to ask about this too, Microsoft doesn't document this function properly so I thought maybe I have to document this myself.

xmine64 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+    }
+
+    /* Lock the registry while we're setting up notifications */
+    CmpLockRegistryExclusive();
```
This is a "change" notification, if someone tries to make a change in a registry key that we want to watch, the other thread/process would try to report the change to an immature NotifyBlock, which is a bad thing. An alternative would be introducing a lock for notifications so when someone wants to report a change they would be blocked until the NotifyBlock is ready.

xmine64 replied:
@GeoB99 I'm waiting for your feedback about this, if the current implementation is ok click on the "Resolve" button and if not let me know so I'll change it.

xmine64 replied:
I'll be going with notification lock, locking KCBs made ReactOS freeze when it's loading a subordinate that's the same as the master key

xmine64 commented:
Sorry, I ran clang-format and I thought I only changed my new code, didn't notice that I also reformatted other codes as well. I'll revert the reformat commit.

xmine64 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+    NTSTATUS Status;
+    KPROCESSOR_MODE PreviousMode = ExGetPreviousMode();
+
+    /* Registry Key */
```
I thought it's a good idea to group variables based on their purpose

xmine64 commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1454,40 +1315,416 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+            /* Probe and capture subordinate objects */
+            if (Count > 0)
+            {
+                LocalSubObjects = ExAllocatePool(NonPagedPool, Count * sizeof(OBJECT_ATTRIBUTES));
```
This a temporary allocation, I'd free them as soon as I'm done with them, I don't think if introducing a new tag for these would be a good idea.

xmine64 replied:
And thanks for `ExAllocatePoolZero`, it made my life easier😃

GeoB99 commented on modules/rostests/apitests/ntdll/NtNotifyChangeMultipleKeys.c:
```diff
@@ -0,0 +1,397 @@
+    /* Make change to the registry key */
+    Status = NtSetValueKey(state->KeyHandle, &ValueName, 0, REG_DWORD, &ValueData1, sizeof(ValueData1));
+    TEST_ERROR(state, Status, STATUS_SUCCESS);
+    SleepEx(100, TRUE);
```
I suppose you're waiting for 0.1 s for the CM to set up notification stuff or why that you set a value to the key two times? Explain thoroughly in comments what's happening here.

xmine64 replied:
I don't know neither😂 I tried different things on Windows and this was the setup that actually worked.
Without the second call, or without the wait, the change is not notified for whatever reason. Could be a synchronization or a timing issue.

xmine64 replied:
Seems like call to Sleep blocks the notification for whatever reason, so instead I used SleepEx with Alertable set to TRUE, hoping that it would work but it didn't neither.

GeoB99 replied:
IMHO mixing up `NtNotifyChangeMultipleKeys` with `NtSetValueKey` and `WaitForSingleObjectEx` together in one single thread execution is a weird way to be notified of changes in a key. Why don't you just split it into two threads, with thread A setting up the notifications and then it waits on the event that the kernel is going to alert you, and thread B toys around with the key's data?

If you want to properly synchronize the threads together so that you get proper test results, you could synchronize your two threads with a dummy event. So basically thread A sets up the notifications, wait on the notification event, thread B sets up new values to the target key and waits on thread A to receive the notification event and then thread A signals the dummy event to the thread B.

xmine64 replied:
I did this for testing the synchronous mode, since it blocks the execution so we won't be able to make changes in registry.
As for asynchronous mode, I think the point of having asynchronous mode is being able to use it in a single-threaded environment, so I think it makes more sense to keep the test for asynchronous mode single threaded.

xmine64 replied:
Maybe I can fix this by increasing watcher thread's priority?

GeoB99 commented on modules/rostests/apitests/ntdll/NtNotifyChangeMultipleKeys.c:
```diff
@@ -0,0 +1,397 @@
+
+    /* Create event */
+    state->EventHandle = CreateEvent(NULL, FALSE, FALSE, NULL);
+    ok(state->EventHandle != NULL, "Failed to create event\n");
```
Failing to create the event handle is a total gameover, what's the point on continuing with the execution of the rest of the test?

GeoB99 commented on modules/rostests/apitests/ntdll/NtNotifyChangeMultipleKeys.c:
```diff
@@ -0,0 +1,397 @@
+    OBJECT_ATTRIBUTES SubordinateObjects[1];
+    InitializeObjectAttributes(&SubordinateObjects[0], &state->SecondaryKeyName, OBJ_CASE_INSENSITIVE, NULL, NULL);
+    Status = NtNotifyChangeMultipleKeys(state->KeyHandle,
+                                        sizeof(SubordinateObjects)/sizeof(OBJECT_ATTRIBUTES),
```
Use `_countof`, `ARRAYSIZE` or whatever

xmine64 replied:
Yup, thanks for the suggestion 
I've forgotten that ARRAYSIZE exists too

xmine64 commented:
The test has been updated.
* Some minor changes requested by GeoB99
* A new test-case that should fail if the same key as master key is provided for subordinate key

Result on Windows 11:
![image](https://github.com/user-attachments/assets/c77367a4-fcba-4031-b76e-bc88211adcc3)

Result on Windows Server 2003:
![image](https://github.com/user-attachments/assets/b4008730-a24e-4da7-a412-f28efbb10509)

Result on ReactOS:
Right now it makes the system freeze. I'll investigate this and update when I fixed it.
![image](https://github.com/user-attachments/assets/96fd05b8-15cc-42f0-ac57-dd133a8263b1)

xmine64 commented:
The new test now works on ReactOS too.

The issue was related to locking KCBs, Now I lock the KCB after opening subordinate objects because if a subordinate object attribute is describing the master it will cause a dead-lock when it's trying to open the subordinate key.

Also I introduced a lock for notification and I use that lock instead of locking the whole registry, so now notifications are blocked until all of the notify blocks are ready. Windows locks the key hive instead but I thought this is a better approach, and also it doesn't have the limitation of subordinates being in separate hives.

![image](https://github.com/user-attachments/assets/859aeffb-add5-4a17-818d-f58726559f5b)

katahiromz commented:
Our `SleepEx` has bug currently:

https://github.com/reactos/reactos/blob/afe29ee7fab548b3431b3fd469b7558c9823f782/dll/win32/shell32/shelldesktop/CDirectoryWatcher.cpp#L39

xmine64 commented:
> Our SleepEx has bug currently
Thanks, my new test doesn't use SleepEx anymore, and the issue wasn't related to SleepEx neither.

xmine64 commented:
I added more test cases that made me discover some more issues. Also I think I need to update KMTEST as well. I'll re-open the PR as soon as I fixed these issues.

xmine64 commented:
I rebased the branch on master to resolve merge conflicts.

@GeoB99 I already tested this branch against my APITEST and KMTEST, and also different software applications (including a SSH server, winbtrfs, Microsoft Office and Safari), I catched and fixed some bugs during these tests and I think it should be stable now, but there might be something I missed or would show itself after a long run.

GeoB99 commented:
@xmine64 Hey sorry for the wait, my notifications are being pilled up on top of yours and I couldn't see it.

```
Also: Should I make a separate PR for APITEST and KMTEST?
```
Yeah, go for it!

tkreuzer commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1452,23 +1452,516 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+    KeEnterCriticalRegion();
+
+    /* Block registry notifications */
+    CmpLockNotifyExclusive();
```
The lock should probably be acquired after capturing the parameters.
Also is there a reason to put `KeEnterCriticalRegion();` here instead of in `CmpLockNotifyExclusive();`?

tkreuzer commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1452,23 +1452,516 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+        {
+            /* Open a kernel-mode handle, since we're using it internally without returning it to the user */
+            HANDLE SubHandle;
+            LocalSubObjects[i].Attributes |= OBJ_KERNEL_HANDLE;
```
This is never allocated, if the call came from kernel mode. What you can do is:
```
    if (PreviousMode != KernelMode)
    {
        // Allocate / probe / capture
        SubordinateObjects = LocalSubObjects;
    }
```
And subsequently only use `SubordinateObjects`. Similar for the `_Out_` parameters.

tkreuzer commented on ntoskrnl/config/ntapi.c:
```diff
@@ -1452,23 +1452,516 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+    if (PostBlock)
+        ExFreePoolWithTag(PostBlock, TAG_CM);
+
+Failure:
```
This is a lot of cleanup code, multiple exit paths, very hard to follow, if it is all correct. A sign, that you should refactor this function and split the inner logic into `CmpNotifyChangeKeys` or something like that.

MuerteSeguraZ commented:
I'll get to reading then
